### PR TITLE
Delete Items

### LINKED ISSUE #12: 11. As a user, I want to be able to delete items from my shopping list so that my list isn’t cluttered with items I don’t want to buy in the future.
It’s possible our users may make an error when entering an item or may decide they won’t be buying a certain item again in the future. Let’s help them keep their lists tidy!

AC:

* User is able to delete an item from the shopping list
* Before deleting, prompt the user to confirm that they really want to delete the item to prevent accidental deletions
* Deletion should cause the associated record(s) in the database to be deleted

### PR DESCRIPTION
**Pull request by kkatoch89**
## Description

This feature will allow for user to delete items from shopping list. Users may make an error when entering an item or may decide they won't be buying a certain item again in the future; this feature will help user to curate their shopping list. Clicking the delete button will prompt user to confirm if user would like to delete in order to prevent accidental deletions.

## Related Issue

closes #12 

## Acceptance Criteria

* User is able to delete an item from the shopping list
* Before deleting, prompt the user to confirm that they really want to delete the item to prevent accidental deletions
* Deletion should cause the associated record(s) in the database to be deleted

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|   ✓ | :link: Update dependencies |
|    | :scroll: Docs              |

### Before

before12

### After

after-delete

## Testing Steps / QA Criteria
To test the feature, use one of the following ways:

1) Follow the link to see the preview: https://tcl-22-shopping-list--pr42-ab-kk-delete-items-roddna01.web.app
2) Launch the app locally from your terminal:
* Pull down this branch with `git pull origin ab-kk-delete-items` and check that branch out with `git checkout origin ab-kk-delete-items`
* Run `npm start`  to launch the app in your browser at [http://localhost:3000](http://localhost:3000)
* A good token to test this feature out on local is `jut testy unix`

When on application:

* In Grocery List view, click on the buttons marked 'X' on the right of each shopping item
* This should display a modal prompting user if they're sure they want to delete the item or not
* If the user clicks the 'Cancel' button, the modal will disappear and the item will remain on the list in both the UI and firestore
* If the user clicks the 'Yes' button, the modal will disappear and the the item will be removed from both the UI and firestore
* [Firebase](https://console.firebase.google.com/u/2/project/tcl-22-shopping-list/firestore/data~2Frosa%20chopin%20tried~2FOp4CU2XUCMAweHISbwxh)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,7 +11,8 @@
     "react-dom": "^17.0.1",
     "react-firebase-hooks": "^3.0.4",
     "react-router-dom": "^5.2.0",
-    "react-scripts": "4.0.1"
+    "react-scripts": "4.0.1",
+    "styled-components": "^5.3.0"
```
Comment (laurenmbeatty) [inline package.json:15]: 🙌

File: src/Item.js [hunk 2]
```diff
@@ -2,10 +2,48 @@ import React, { useState, useEffect } from 'react';
...
+  color: black;
+  text-align: center;
+  cursor: pointer;
+  background: lightgray;
+  justify-content: center;
+  border: none;
+  &:hover {
+    background: darkgray;
+  }
+`;
```
Comment (dariatsvetkova) [inline src/Item.js:42]: Love the use of styled-components - makes the code really easy to read!

File: src/Item.js [hunk 3]
```diff
@@ -2,10 +2,48 @@ import React, { useState, useEffect } from 'react';
...
 
 function Item({ userToken, item }) {
```
Comment (dariatsvetkova) [inline src/Item.js:43]: At this point, most of the code in this file is for the modal and its functionality (deleting items from the database). Would it make sense to extract the modal into its own component?

**Review by dariatsvetkova — COMMENTED** [review]
Great work, team! It is starting to look like a real app, thanks to the modal :) I just have one comment that I'd like to discuss.

**Comment by cafloyd** [comment]
https://github.com/the-collab-lab/tcl-22-smart-shopping-list/pull/41#issuecomment-840786930
See comment above from @laurenmbeatty !

File: src/Item.js (hunk 3 above)
Comment (cafloyd) [reply src/Item.js:43]: I agree that this is worth considering!

File: src/Item.js [hunk 4]
```diff
@@ -71,6 +109,27 @@ function Item({ userToken, item }) {
...
+        console.log('it worked!');
+      });
+  };
+
+  // [delete-item] 4b. If no, then the modal closes
+  const handleCloseModal = () => {
+    setOpenModal(false);
+  };
+
+  // [delete-item] 3. Create a function that is passed to onClick which will render a mondal which asks the user 'You sure about dat?'
```
Comment (cafloyd) [inline src/Item.js:128]: 😂

**Review by cafloyd — APPROVED** [review]
Love it! Such nice implementation of a very useful feature! I think Lauren's suggestion of having the item name in the popup modal could make for even smoother UX if we want to implement it now or in the coming weeks!

**Comment by laurenmbeatty** [comment]
Oh, I see you're using material-ui. I'm surprised I can't tab to the "buttons". Is there a reason you're using `a` tags instead of actual buttons? If you're styling something to look like a button, then it should be a button. Here's some accessibility info from material-ui.
image

File: src/Item.js [hunk 5]
```diff
@@ -2,10 +2,48 @@ import React, { useState, useEffect } from 'react';
...
+  box-shadow: 0 3rem 5rem rgba(0, 0, 0, 0.3);
+  z-index: 10;
+  max-width: 250px;
+  font-size: 15px;
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+`;
+
+const ModalBtn = styled.a`
```
Comment (laurenmbeatty) [inline src/Item.js:26]: I would make these actual buttons instead.

File: src/Item.js [hunk 6]
```diff
@@ -71,6 +109,27 @@ function Item({ userToken, item }) {
...
   };
 
+  // [delete-item] 4a. If yes, then remove from firestore database
+  const handleDelete = () => {
+    setOpenModal(true);
+    db.collection(userToken)
+      .doc(id)
+      .delete()
+      .then(() => {
+        console.log('it worked!');
```
Comment (laurenmbeatty) [inline src/Item.js:119]: Could use your snackbar that's already installed to send a success message.

File: src/Item.js [hunk 7]
```diff
@@ -83,6 +142,25 @@ function Item({ userToken, item }) {
...
+      {/* // [delete-item] 2. Create a button modal that will render when user tries to delete item */}
+      <Modal
+        open={openModal}
+        onClose={handleCloseModal}
+        aria-labelledby="simple-modal-title"
+        aria-describedby="simple-modal-description"
+      >
+        <div>
+          <StyledModal>
+            <h1>Are you sure you want to delete this item?</h1>
```
Comment (laurenmbeatty) [inline src/Item.js:158]: Looks like you need to add an `id="simple-modal-title"` to complete the `aria-labelledby` connection.

File: src/Item.js [hunk 8]
```diff
@@ -83,6 +142,25 @@ function Item({ userToken, item }) {
...
+      {/* // [delete-item] 1. Create a button next to each item in list */}
+      <button type="button" onClick={handleOpenModal}>
+        X
+      </button>
+      {/* // [delete-item] 2. Create a button modal that will render when user tries to delete item */}
+      <Modal
+        open={openModal}
+        onClose={handleCloseModal}
+        aria-labelledby="simple-modal-title"
+        aria-describedby="simple-modal-description"
```
Comment (laurenmbeatty) [inline src/Item.js:154]: You don't have a modal description, so doesn't look like you need the `aria-describedby`

File: src/Item.js (hunk 3 above)
Comment (ablizben) [reply src/Item.js:43]: This is true, I would think we could do that. If Karan agrees, we could go ahead and move it to a separate component.

File: src/Item.js (hunk 4 above)
Comment (ablizben) [reply src/Item.js:128]: I forgot that comment was still in there 😂.

File: src/Item.js (hunk 3 above)
Comment (kkatoch89) [reply src/Item.js:43]: For sure! That's a great point!

File: src/Item.js (hunk 7 above)
Comment (kkatoch89) [reply src/Item.js]: Oh good point!

File: src/Item.js (hunk 6 above)
Comment (kkatoch89) [reply src/Item.js]: Great point!

File: src/Item.js [hunk 9]
```diff
@@ -71,6 +77,34 @@ function Item({ userToken, item }) {
...
+      .delete()
+      .then(() => {
+        enqueueSnackbar(`${itemName} successfully deleted`, {
+          variant: 'warning',
+        });
+      })
+      .catch(() => {
+        enqueueSnackbar('Oops! Something went wrong.', {
+          variant: 'error',
+        });
```
Comment (laurenmbeatty) [inline src/Item.js:94]: Perfect!

File: src/StyledModal.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
+import styled from 'styled-components';
+
+export const StyledModalBtn = styled.button`
```
Comment (laurenmbeatty) [inline src/StyledModal.js:3]: This fixed the accessibility issue. 💯

File: src/Item.js [hunk 11]
```diff
@@ -83,6 +117,29 @@ function Item({ userToken, item }) {
...
+        X
+      </button>
+      {/* // [delete-item] 2. Create a button modal that will render when user tries to delete item */}
+      <Modal
+        open={openModal}
+        onClose={handleCloseModal}
+        aria-labelledby="simple-modal-title"
+      >
+        <StyledModalLayout>
+          <h1>
```
Comment (laurenmbeatty) [inline src/Item.js:131]:
```suggestion
          <h1 id="simple-modal-title">
```
Probably better would be `delete-modal-title` (same with the `aria-labelledby` above)

**Review by laurenmbeatty — APPROVED** [review]
Looks great!!!! Excited that you all tried `styled-components`

**Review by jonathanyeong — APPROVED** [review]
Love the modal!

**Pull request merged**